Add Param Digger Blog (#1208)
zaproxy/zaproxy-website: site/content/blog/2022-08-22-the-param-miner-addon/index.md

---
title: "The Param Miner Add-on"
summary: >
  The parameter discovery add-on for ZAP.
images:
- https://www.zaproxy.org/blog/2022-08-22-the-param-miner-addon/images/paramminer.png
type: post
tags:
- blog
date: "2022-08-22"
authors: 
    - Arkaprabha
---

> Parameter discovery is the process of finding unlinked, unreferenced, or hidden parameters that can be used by the user to control or influence the behaviour of the application. Resulting in successful attack scenarios like SQL injection, XSS, SSRF, Web Cache Poisoning based attacks and much more. Often times such uncovered parameters are properly configured and can be a suitable entry point for further manipulation. These are real hidden treasures which can be very crucial in the recon process to develop a successful exploit.  

Reviewer: `SQL Injection`, `XSS` etc could all link to the relevant ZAP issues? eg https://www.zaproxy.org/docs/alerts/40018/ ...

> Param Miner is a new add-on for ZAP that allows you to discover and analyse parameters in web applications, that might not be visible to the user. The main idea behind Param-Miner is to increase the attack surface of the website by identifying these parameters.

Reviewer: First "Param Miner" (Digger;) mention could link to the help pages, which need to be put on the Marketplace...

---

> title: "The Param Miner Add-on"

Author: I would add it to the title?

> title: "The Param Miner Add-on"

Reviewer 2: Ya, that’s what I was thinking

> title: "The Param Miner Add-on"

Reviewer: "The New Param Digger Add-on"?

summary: >
  The parameter discovery add-on for ZAP.
images:

> - https://www.zaproxy.org/blog/2022-08-22-the-param-miner-addon/images/paramminer.png

Reviewer 2: @psiinon do you think you can knock together a fun social preview image with zapbot & and a [mining pick](https://www.google.ca/search?q=mining+pick)?

Author: We had the mining pick at the beginning actually :P It was then changed to hard-mine-helmet! :) So our logo now is a mining helmet...

Reviewer 2: Okay zapbot with a mining helmet :grinning:

Author: Looks like we need to change to a "mole" ... Sort of like the icon for golang :)... The pokemon Digglet would be good too :P

Reviewer: I'll see what I can do... :)

type: post
tags:

> - blog

Author: I feel like we should :P but :)

date: "2022-08-22"
authors: 
    - Arkaprabha
---

> Parameter discovery is the process of finding unlinked, unrferenced or hidden parameters that can be used by the user to control the behaviour of the application. These can allow an attacker to control the behaviour of the application resulting in succesful attack scenarios like SQL injection, XSS, SSRF, Web Cache Poisoning based attacks and much more. Often times these are properly configured and can be a suitable entrypoint. These are real hidden treasures which can be very crucial in the recon process to develop a successful exploit.  

Reviewer 2: I dunno, it still feel a bit repetitive or something, but I think that's better.

> The Param Miner Add-on is a new add-on for ZAP that allows you to discover and analyse parameters in web applications, that might not be visisble to the user. The main idea behind Param-Miner is to increase the attack surface of the website by identifying these parameters. 

Reviewer 2: ```suggestion
Param Miner is a new add-on for ZAP that allows you to discover and analyse parameters in web applications, that might not be visible to the user. The main idea behind Param-Miner is to increase the attack surface of the website by identifying these parameters. 
```

Need to be consistent on `Param Miner` or `Param-Miner` **throughout** the doc

Author: Twitter might pass judgement on us anyway :P but Param Miner it is :)

The add-on is still in its early stage, so the range of functionality is limited. I'll be adding more functionality over the next few days and in the long term I hope to add it to the Automation Framework in ZAP.

This post highlights the available features of the Param-Miner add-on (as of 2022-08-22).

## How to use the Param-Miner add-on

> To begin with, you will need install it from the [Marketplace](/addons/) from within ZAP as it's not installed by default. Once you have it installed, you can access the Param Miner from three locations: the add-on menu which appears after clicking the plus green icon, the right click menu under the Attack option and from the tools menu. 

Reviewer 2: ```suggestion
To begin with, you will need install it from the [Marketplace](/addons/) from within ZAP as it's not installed by default. Once you have it installed, you can access the Param Miner from three locations: the add-on menu which appears after clicking the plus green icon, the right click menu under the Attack option, and from the tools menu. 
```

![Right click menu](images/right-click-menu-example.png)

This is how the Param Miner panel in the information window looks like:
![Param Miner Tab](images/paramminer.png)

> Whether you choose to start a new scan by clicking the "New Scan" button on the Param Miner tab or start a new scan from the attack menu or the tools menu, you will be greeted with the Param Miner dialog. This dialog allows you to select the parameters you want to discover, the methods you want to use, the word lists you want to use and other options.  

Reviewer 2: ```suggestion
Whether you choose to start a new scan by clicking the "New Scan" button on the Param Miner tab or start a new scan from the attack menu or the tools menu, you will be greeted with the Param Miner dialog. This dialog allows you to select the parameters you want to discover, the methods you want to use, the word lists you want to use, and other options.  
```

![Param Miner Dialog](images/param-miner-dialog.png)

> As of 2022-08-20 the Param Miner only supports the "URL Guess", which uses a list of URL parameters. So it is the one which is selected by default. The screenshot of the default "URL Guess" tab is given below. You can use the predefined wordlist provided by the add-on or you can create your own and import it after selecting the "Custom" option from the Wordlist pull down. Selecting "Custom" would also enable the File Location field for you where you can supply the appropriate file path after browsing. You can also use both the wordlists provided by the add-on and your own by selecting the "Both" option. 

Reviewer 2: Use the 22nd for consistency?

Author: yep!

> 

Reviewer 2: Drop one blank line

After selection of the word list, you have to select the methods. By default it's GET. You can select as many methods as you want using Ctrl + Click (For Mac or other related platforms use Cmd + Click). 

![URL Guess Tab](images/url-guess-tab.png)

> 

Reviewer 2: Drop one blank line

## Future Work

> While the foundation of the add-on has been laid out, the Param Miner is still in its early stage. There is a lot more work left to do. I would be adding more functionality such as a starter string for JSON method in URL Guesser, adding a new method for XML and other guessing methods in the Param Miner dialog. I would also be adding "Parameter Mining" using "FAT GET" requests and other techniques as suggested in [Practical Web Cache Posinoning](https://portswigger.net/research/practical-web-cache-poisoning) and [Web Cache Entanglement](https://portswigger.net/research/web-cache-entanglement) which can be enabled or disabled based on users' choice, having the name "web cache poisoning" or other such related term.

Reviewer 2: ```suggestion
While the foundation of the add-on has been laid out, the Param Miner is still in its early stage. There is a lot more work left to do. I will be adding more functionality such as: A starter string for the JSON method in the URL Guesser, adding a new method for XML, and other guessing methods in the main dialog. I will also be adding "Parameter Mining" using "FAT GET" requests and other techniques as suggested in [Practical Web Cache Posinoning](https://portswigger.net/research/practical-web-cache-poisoning) and [Web Cache Entanglement](https://portswigger.net/research/web-cache-entanglement) which can be enabled or disabled based on users' choice, having the name "web cache poisoning" or other such related term.
```

> I would also work on add-on's integration with the Automation Framework in ZAP. This would help users to find parameters on every URLs in the website by using the Spider and launch active or passive scans against the results.

Reviewer 2: ```suggestion
I would also like to work on the add-on's integration with the ZAP Automation Framework. This would help users to find parameters on every URL in the website by using the Spider and launch active or passive scans against the results.
```

Have a play with this add-on and if you can think of anything it could do better then get in touch or open issues on Github!  

> [google groups](https://groups.google.com/group/zaproxy-users)

Reviewer 2: ```suggestion
[User Group](https://groups.google.com/group/zaproxy-users)
```
